### ISSUE TITLE
template_file render not waiting on dependencies

### ISSUE THREAD

**Issue by mengesb** [hashicorp/terraform#5773]
I've been trying to narrow down an issue I'm having sourcing plan generated files via `template_file`.

The approach I've taken is to submodule out the generation of dynamic files I need to read, and then use that module's 'output' method to use the file in the parent plan. The module outputs the contents and path to the required file which I then use in the template `vars` segment.

Example:
- need to create a user on a server, and source their private key to create a chef data bag.

Versions attempted:
- `0.6.13`
- `0.6.14`

template_file:

```
{
  "id": "delivery",
  "delivery_pem": "${key2}"
}
```

template_file invocation:

```
resource "template_file" "delivery-json" {
  template = "${file("${path.module}/files/delivery-json.tpl")}"
  vars {
    key2 = "${module.delivery-user.pem}"
  }
}
```

module delivery-user invocation:

```
module "delivery-user" {
  source      = "delivery-user"
  fqdn        = "${var.chef_dns}"
  private_key = "${var.aws_private_key_file}"
  server_user = "${lookup(var.ami_usermap, var.ami_os)}"
  org         = "${var.chef_org}"
}
```

delivery-user module:

```
variable "fqdn" {
  description = "Chef server fqdn"
}
variable "private_key" {
  description = "Full path to your local private key"
}
variable "server_user" {
  description = "Default username to ssh into Chef Server with"
  default     = "centos"
}
variable "username" {
  description = "Username of the first Delivery user"
  default     = "delivery"
}
variable "user_firstname" {
  description = "Delivery user first name on Chef Server"
  default     = "Delivery"
}
variable "user_lastname" {
  description = "Delivery user last name on Chef Server"
  default     = "User"
}
variable "user_email" {
  description = "Delivery user's e-mail address"
  default = "delivery@domain.tld"
}
variable "org" {
  description = "Chef Server organization short name (lowercase alphanumeric characters only)"
}
resource "null_resource" "delivery-user" {
  connection {
    user        = "${var.server_user}"
    private_key = "${var.private_key}"
    host        = "${var.fqdn}"
  }
  # Local directory setup
  provisioner "local-exec" {
    command = "rm -rf .delivery/${username}.pem ; mkdir -p .delivery"
  }
  # Delete possible ${var.username}
  provisioner "remote-exec" {
    inline = [
      "rm -rf .delivery/${username}.pem",
      "sudo chef-server-ctl user-delete ${var.username} -y || echo Deleted ${var.username}",
    ]
  }
  # Create ${var.username}
  provisioner "remote-exec" {
    inline = [
      "mkdir -p .delivery",
      "sudo chef-server-ctl user-create ${var.username} ${var.user_firstname} ${var.user_lastname} ${var.user_email} ${base64encode(self.id)} -f .delivery/${var.username}.pem",
      "sudo chef-server-ctl org-user-add ${var.org} ${var.username} -a",
      "sudo chown ${var.server_user} .delivery/${var.username}.pem",
    ]
  }
  # Harvest .delivery/${username}.pem
  provisioner "local-exec" {
    command = "scp -o StrictHostKeyChecking=no -i ${var.private_key} ${var.server_user}@${var.fqdn}:.delivery/${var.username}.pem .delivery/${var.username}.pem"
  }
  # Chef Server cleanup
  provisioner "remote-exec" {
    inline = [
      "rm -rf .delivery/${username}.pem",
    ]
  }
  provisioner "local-exec" {
    command = "cat .delivery/${var.username}.pem"
  }
}
output "pem" {
  value = "${file(".delivery/${var.username}.pem")}"
}
output "pem_file" {
  value = ".delivery/${var.username}.pem"
}
```

debugging:

```
template_file.delivery-json: Creating...
  rendered: "" => "<computed>"
  template: "" => "{\n  \"id\": \"delivery\",\n  \"delivery_pem\": \"${key2}\"\n}\n"
  vars.#:   "" => "<computed>"
module.ssh-keygen.null_resource.ssh-keygen: Provisioning with 'local-exec'...
module.delivery-user.null_resource.delivery-user: Creating...
...
module.delivery-user.null_resource.delivery-user: Provisioning with 'local-exec'...
module.ssh-keygen.null_resource.ssh-keygen (local-exec): Executing: /bin/sh -c "rm -rf .delivery/builder_key* ; mkdir -p .delivery"
...
Error applying plan:

2 error(s) occurred:

* template_file.delivery-json: failed to render : 3:23: unknown variable accessed: key2
* local-exec provisioner missing 'command'

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

Now... if there was something in `module` where I could `depends_on` then maybe the module would render fully before we move on... but when I look at the dependency graph I see no container for the submodules either ... so I'm wondering if there's something broken here with 0.6.x? What is displayed above is the submodule processing, but the plan moves on without regard to the requirement. Looks like the `${module.delivery-user.pem}` output requirement is being completely ignored and it finishes the `template_file`, thus breaking the render.

**Comment by rantecki** [comment]
+1 Just ran into the same issue.

**Comment by apparentlymart** [comment]
Hi @mengesb and @rantecki,

This looks like a pretty weird dependency handling bug. I'm curious though about whether rendering the template using the data source mechanism introduced in 0.7 would work around this issue.

To do this, you could change from `resource "template_file"` to `data "template_file"`, and update the referencing interpolations to be `${data.template_file.delivery-json.rendered}` instead of just `${template_file.delivery-json.rendered}` .

In this situation the use of a data resource might not actually address the problem, because the data resource will still depend on that module and thus its resolution will be delayed until the "apply" phase. But if it does happen to work, it'd hopefully be a reasonable workaround in the short term and might help us narrow down what part of Terraform is buggy here.

**Comment by rantecki** [comment]
@apparentlymart actually it looks like I was already using the data syntax.

Simplified scenario below:

```
module "ecr_repository" {
    source = "../image-repository"
}

data "template_file" "gameserver" {
  template = "${file("./aws-task-definitions/gameserver.json")}"
  vars {
    gameserver_image = "${module.ecr_repository.repository_url}"
  }
}

resource "aws_ecs_task_definition" "gameserver" {
  family = "gameserver"
  container_definitions = "${data.template_file.gameserver.rendered}"
}
```

In this case I'd expect the `ecr_repository` module to applied before `aws_ecs_task_definition`.

plan output:

```
+ aws_ecs_task_definition.gameserver
    arn:                   "<computed>"
    container_definitions: "529c7dbc93034824d511c31b185d687350476127"
    family:                "gameserver"
    network_mode:          "<computed>"
    revision:              "<computed>"

<= data.template_file.gameserver
    rendered: "<computed>"
    template: "...{snipped}..."
    vars.%:   "<computed>"

+ module.ecr_repository.aws_ecr_repository.gameserver
    arn:            "<computed>"
    name:           "gameserver"
    registry_id:    "<computed>"
    repository_url: "<computed>"

Plan: 2 to add, 0 to change, 0 to destroy.
```

The module is applied last.

Interesting to note though, that the graph output appears to show the correct dependency relationship.

![graph2](https://cloud.githubusercontent.com/assets/1551714/19330219/dd820c38-9128-11e6-9dd3-db854e1be6da.png)

**Comment by apparentlymart** [comment]
Thanks for these additional details, @rantecki!

I tried to cut this down a bit further so that I could test it without the AWS provider in the mix.

Here's my root module:

```
module "ecr_repository" {
    source = "./image-repository"
}

data "template_file" "gameserver" {
  template = "${file("./gameserver.json")}"
  vars {
    gameserver_image = "${module.ecr_repository.repository_url}"
  }
}

resource "null_resource" "woot" {
  triggers {
    family = "gameserver"
    container_definitions = "${data.template_file.gameserver.rendered}"
  }
}
```

Here's my `./image-repository` module:

```
resource "template_file" "hi" {
  template = "http://example.com/"

  provisioner "local-exec" {
    command = "sleep 5"
  }
}

output "repository_url" {
    value = "${template_file.hi.rendered}"
}
```

Here I'm using a `template_file` resource, doing it the old-fashioned 0.6 way as a managed resource so that I can get a "computed" attribute to stand in for your computed repository URL. I used a `local-exec` provisioner to "encourage" it to complete after the other resources if possible.

Here's the plan:

```
<= data.template_file.gameserver
    rendered: "<computed>"
    template: "{\n    \"image\": \"${gameserver_image}\"\n}\n"
    vars.%:   "<computed>"

+ null_resource.woot
    triggers.%: "<computed>"

+ module.ecr_repository.template_file.hi
    rendered: "<computed>"
    template: "http://example.com/"
```

Here's the result of applying this plan:

```
module.ecr_repository.template_file.hi: Creating...
  rendered: "" => "<computed>"
  template: "" => "http://example.com/"
module.ecr_repository.template_file.hi: Provisioning with 'local-exec'...
module.ecr_repository.template_file.hi (local-exec): Executing: /bin/sh -c "sleep 5"
module.ecr_repository.template_file.hi: Creation complete
data.template_file.gameserver: Refreshing state...
null_resource.woot: Creating...
  triggers.%:                     "" => "2"
  triggers.container_definitions: "" => "{\n    \"image\": \"http://example.com/\"\n}\n"
  triggers.family:                "" => "gameserver"
null_resource.woot: Creation complete

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

So for my config here it seems like Terraform processed the resources in the expected order, with the child module's resource going first.

Do you see anything in my example (apart from the obvious lack of ECR resources) that differs from yours?

**Comment by rantecki** [comment]
Thanks for the example, it worked fine.  I admit I hadn't actually applied my original configuration as the order of things didn't look right, so I didn't even attempt it.  But lo and behold, I just tried running my original scenario and it worked!

This may be a misunderstanding on my behalf as I assumed that the plan output was in the order that things will occur when applied, but this does not seem to be the case?

So it looks like my case at least is a non-issue.  I'm not sure about the original poster's issue.

**Comment by apparentlymart** [comment]
Ahh yes @rantecki, I think that explains it... the plan output is sorted into lexicographical order for easy reference, and isn't the order in which the steps will be applied.

**Comment by chadgrant** [comment]
@apparentlymart so I am using a lot of includes via: -var-file=x when I use plan, but I do not include these when I run apply plan-file. 

From what it seems you said in this thread that some variables/templates may not calculate until the apply phase?

if so ... I would have to pass the var-files along to apply as well....

**Comment by andresvia** [comment]
@apparentlymart 

I'm having this issue with `0.10.6`.

During TF plan, everything is ok. During apply, I'm getting the `failed to render unknown variable accessed`

Note the var failed to access is a `${var.something}`

```
<= module.ecs_service.data.template_file.task_definition
      id:                      <computed>
      rendered:                <computed>
      template:                "...\"${image_url}\"..."
      vars.%:                  <computed>
```

The template `vars{}` section contains all sort of things, variables, module outputs, data sources, resources, locals, `"${path.module}"`, one var has a ternary with a file. 

```
vars {
...
  foo = "${file(var.custom_foo_file == "" ? "${path.module}/default.foo.file" : var.custom_foo_file)}"
...
}
```

The template is itself contains a ternary with a file.

```
data "template_file" "bar" {
  template = "${file(var.custom_bar_template == "" ? "${path.module}/default.bar.template" : var.custom_bar_template)}"
...
   vars {
   ...
     foo = "${file(var.custom_foo_file == "" ? "${path.module}/default.foo.file" : var.custom_foo_file)}"
   ...
   }
}
```

Oh, and one of the template vars is coming from a module that uses an external data source.

**Comment by andresvia** [comment]
After debugging found that this happened when the external data source mentioned above returned an empty object `{}` instead of failing when the external resource didn't exist, we used this as a convention for "the resource doesn't exist, just return an empty value", to be clear we were not experiencing dependency issues, but instead not following the pattern of other data sources, to fail when the resource doesn't exist.

**Comment by teamterraform** [comment]
Hi all! Sorry for the long silence.

A lot has changed since Terraform 0.6, not least of which the recent release of Terraform 0.12 which includes a new configuration language engine. As a consequence, it's not really possible to map what's described above into the current implementation, though we've seen similar configurations to those shown here working just fine in recent versions of Terraform.

Although it's unsatisfying to close an issue without a definite solution, we're going to optimistically close this expecting that it was fixed as part of Terraform 0.12, if not before. If anyone is still seeing errors that relate to dependencies not being respected, please do open a _new_ issue for it so we can capture a fresh set of information to test and debug with, in the context of the new codepaths in Terraform 0.12.
